### DISCUSSION TITLE
Accessing numba function directly

### DISCUSSION THREAD

**Discussion by aseyboldt** [aesara-devs/aesara#1009] [General]
I can see quite a few use cases for accessing the generated numba functions directly, instead of using them through the aesara function wrapper. Mostly if we want to create a function pointer that we can call from other compiled code, using `numba.cfunc`.

I'm using this is a nuts sampler implemented in rust, and it would also be very useful for sunode, where we could compile the right hand side of an ODE as aesara function, compile that and pass a pointer to the function to the c library sundials to do the actual work.

In nutpie, I currently compile a function in numba mode, then access the optimized fgraph and then use numba_funcify to get an optimized numba function:
https://github.com/aseyboldt/nutpie/blob/master/nutpie/compile_pymc.py#L124

This works, but seems like a rather roundabout way of doing it. I also think it ends up compiling the numba functions twice, and thus slows things down a lot.

Did anyone else do something like this already? Should I just call the the optimization code myself, or should there be an interface for it?

**Comment by brandonwillard** [comment]
The JIT compiled functions produced by `JITLinker` are made available through [a `.jit_fn` attribute](https://github.com/aesara-devs/aesara/blob/d6858fe23af29cb28379723175895c9d02fdca51/aesara/link/basic.py#L732).  More specifically, the Numba/JAX compilation result is available as `fn.vm.jit_fn` for an `aesara.function`-compiled function `fn`.

**Reply by brandonwillard** [reply]
Also, here's an example where the individual node functions are obtained (and printed): https://github.com/aesara-devs/aesara/issues/923#issuecomment-1126630277.  A similar approach could be used to get much more granular JIT-compiled parts of a graph.

Really, all this involves an intermediate form consisting of plain Python functions that are JIT compiled by external packages.  Those intermediate forms currently aren't available directly, but it's not impossible to obtain them from the high-level compilation results.  Regardless, it wouldn't be difficult to produce a `Linker` subclass that tracks this information in a more accessible way.

**Reply by aseyboldt** [reply]
Ah, thanks I missed that.

Getting our hands at intermediate results would be pretty nice for debugging performance or correctness bugs as well I think. This would allow us to look at the generated numba functions and see what it is doing.

**Reply by aseyboldt** [reply]
This does speed up the compilation, but it still seems rather slower than I think it should be. If for example I take a model like this:

```python
with pm.Model(coords=coords, check_bounds=False) as pymc_model:
    intercept = pm.Normal("intercept", sigma=10)

    # County effects
    # TODO should be a CenteredNormal
    raw = pm.Normal("county_raw", dims="county")
    sd = pm.HalfNormal("county_sd")
    county_effect = pm.Deterministic("county_effect", raw * sd, dims="county")

    # Global floor effect
    floor_effect = pm.Normal("floor_effect", sigma=2)

    # County:floor interaction
    # Should also be a CenteredNormal
    raw = pm.Normal("county_floor_raw", dims="county")
    sd = pm.HalfNormal("county_floor_sd")
    county_floor_effect = pm.Deterministic(
        "county_floor_effect", raw * sd, dims="county"
    )

    mu = (
        intercept
        + county_effect[county_idx]
        + floor_effect * data.floor.values
        + county_floor_effect[county_idx] * data.floor.values
    )

    sigma = pm.HalfNormal("sigma", sigma=1.5)
    pm.Normal(
        "log_radon", mu=mu, sigma=sigma, observed=data.log_radon.values, dims="obs_id"
    )
```

And get a logp function (even without the gradient)
```
func = aesara.function(
    model.value_vars, model.logpt(), mode=aesara.compile.NUMBA
)
```

This takes about 2s, most of that time in `compiler_machinery.py:264(_runPass)`. But why is it even compiling things already? Shouldn't this happen lazily if we call it the first time, or include it in a `numba.cfunc`?

For my cfunc I still end up with ~13s in total for the dlogp function...

**Reply by aseyboldt** [reply]
It's just that this leads to lots of models where the timings look something like this:
![image](https://user-images.githubusercontent.com/1882397/175397865-40a04940-7941-4641-b9f4-2f86b7b554f3.png)

**Reply by ricardoV94** [reply]
> It's just that this leads to lots of models where the timings look something like this:
> ![image](https://user-images.githubusercontent.com/1882397/175397865-40a04940-7941-4641-b9f4-2f86b7b554f3.png)
> 

What numbers should "jump" there?

**Reply by aseyboldt** [reply]
That compilation takes 20 times longer than sampling. After all that work to get sampling fast in nutpie, we now have to wait for the compiler :D

**Reply by brandonwillard** [reply]
Ah, yes, some transpilation dispatch functions pre-compile parts of the subgraphs they produce when they provide explicit type signatures to `numba.njit`.  This is/was necessary for some uses of the `numba.vectorize` machinery, I believe&mdash;and perhaps something else I'm forgetting.

**Reply by ricardoV94** [reply]
> That compilation takes 20 times longer than sampling. After all that work to get sampling fast in nutpie, we now have to wait for the compiler :D

Any idea where the biggest slowdowns come from? Does that compile include the whole pipeline pymc->aeppl->aesara optim->numba compilation?

**Reply by brandonwillard** [reply]
> That compilation takes 20 times longer than sampling. After all that work to get sampling fast in nutpie, we now have to wait for the compiler :D

If you're talking about times at such small scales, then I don't think that really matters.  ~Regardless, if you want to reduce the compilation time, disable optimizations and other things that aren't necessary (i.e. simply perform the transpilation to Numba).~

Wait, you're talking about a _different_ form of compilation (i.e. whatever's happening in `nutpi.compile_pymc_model`)?  If that's the case, I can't really say anything without more information; otherwise, if compilation only occurs once (for a given graph), then these results are really good.

**Reply by aseyboldt** [reply]
@ricardoV94 It includes the whole pipeline, but most of the time is spent in numba functions (again the `compiler_machinery.py:264(_runPass)`). See profile below.

@brandonwillard Disabling some optimizations is certainly something that I'll try. I think those times matter a lot in practice though. If you iterate a lot on a model you go through the modify -> compile -> sample -> modify cycle quite a lot of times, and it feels very different if that takes 20s vs 5s or so.
And I'm not sure we are anywhere close to the optimum here, your comment about the jit_fn argument already saved 5s. :-)
A couple more of that and things are getting really nice...

![image](https://user-images.githubusercontent.com/1882397/175399863-be86a872-4c25-42aa-a9e0-d7c824284673.png)

**Reply by aseyboldt** [reply]
@brandonwillard Yes, it gets embedded into something else, namely a `numba.cfunc` (https://github.com/aseyboldt/nutpie/blob/master/nutpie/compile_pymc.py#L169).
This is why I would like to avoid any previous compilation steps, because (I think) those will not be used when the cfunc is compiled.

**Reply by brandonwillard** [reply]
`compiler_machinery` is a component of the Numba compilation pipeline, so the next step would involve determining when/where/how often Numba compilation is occuring, and then whether or not any of that is performed redundantly due to the Aesara compilation/transpilation pipeline.

The Aesara compilation/transpilation pipeline shouldn't be calling any of the dispatch functions that perform intermediate Numba compilations, or the final, entire-graph compilations, more than once for a given node/graph, so, in a sense, this compilation latency should be close to the latency one would see when compiling the same graph by hand with Numba.

As I mentioned previously, we are adding some additional compilation passes/settings to the standard Numba compilation pipeline (e.g. in order to get performant vectorized operations in certain cases), so keep that in mind.

**Reply by brandonwillard** [reply]
Also, the Numba cache should be enabled by default, but I don't know how much of the graph you're using is covered by Numba's caching.  This is something we haven't considered&mdash;but need to&mdash;regarding the type of Python functions we generate in Aesara for Numba (i.e. we might be producing graphs that can't be adequately cached when we could be, or not caching with enough sub-graph granularity to help).

**Reply by aseyboldt** [reply]
Can you point me to those additional passes? I can't find them...

I played a little bit with the cache argument, but so far I never observed any difference, no clue why.

**Reply by brandonwillard** [reply]
Anywhere where `aesara.link.numba.dispatch.basic.use_optimized_cheap_pass` is used will invoke an extra optimization pass in Numba.  Again, this is all related to performance concerns for element-wise reduction operations in Numba (see https://github.com/aesara-devs/aesara/issues/529).
